src/benchmarklib/tpcds/tpcds_table_generator.cpp

```diff
@@ -0,0 +1,1141 @@
+  auto k_first_row = ds_key_t{};
+
+  split_work(table_id, &k_first_row, &k_row_count);
+
+  const auto& tdefs = *getSimpleTdefsByNumber(table_id);
+
+  if (k_first_row != 1) {
+    row_skip(table_id, static_cast<int>(k_first_row - 1));
+    if (tdefs.flags & FL_PARENT) {
+      row_skip(tdefs.nParam, static_cast<int>(k_first_row - 1));
+    }
+  }
+
+  if (tdefs.flags & FL_SMALL) {
+    resetCountCount();
+  }
+
+  Assert(k_row_count <= std::numeric_limits<tpcds_key_t>::max(),
+         "tpcds_key_t is too small for this scale factor, "
+         "consider using tpcds_key_t = int64_t;");
+  return {k_first_row, k_row_count};
+}
+
+pmr_string boolean_to_string(bool boolean) { return pmr_string(1, boolean ? 'Y' : 'N'); }
+
+pmr_string zip_to_string(int32_t zip) {
+  auto result = pmr_string(5, '?');
+  std::snprintf(result.data(), result.size() + 1, "%05d", zip);
+  return result;
+}
+
+// dsdgen deliberately creates NULL values if nullCheck(column_id) is true, resolve functions mimic that
+std::optional<pmr_string> resolve_date_id(int column_id, ds_key_t date_id) {
+  if (date_id <= 0 || nullCheck(column_id)) {
+    return std::nullopt;
+  }
+
+  auto date = date_t{};
+  jtodt(&date, static_cast<int>(date_id));
+
+  auto result = pmr_string(10, '?');
+  std::snprintf(result.data(), result.size() + 1, "%4d-%02d-%02d", date.year, date.month, date.day);
+
+  return result;
+}
+
+std::optional<tpcds_key_t> resolve_key(int column_id, ds_key_t key) {
+  return nullCheck(column_id) || key == -1 ? std::nullopt : std::optional{static_cast<tpcds_key_t>(key)};
+}
+
+std::optional<pmr_string> resolve_string(int column_id, pmr_string string) {
+  return nullCheck(column_id) || string.empty() ? std::nullopt : std::optional{std::move(string)};
+}
+
+std::optional<int32_t> resolve_integer(int column_id, int value) {
+  return nullCheck(column_id) ? std::nullopt : std::optional{int32_t{value}};
+}
+
+std::optional<float> resolve_decimal(int column_id, decimal_t decimal) {
+  auto result = 0.0;
+  dectof(&result, &decimal);
+  // we have to divide by 10 after dectof to get the expected result
+  return nullCheck(column_id) ? std::nullopt : std::optional{static_cast<float>(result / 10)};
+}
+
+std::optional<float> resolve_gmt_offset(int column_id, int32_t gmt_offset) {
+  return nullCheck(column_id) ? std::nullopt : std::optional{static_cast<float>(gmt_offset)};
+}
+
+std::optional<pmr_string> resolve_street_name(int column_id, const ds_addr_t& address) {
+  return nullCheck(column_id) ? std::nullopt
+                              : address.street_name2 == nullptr
+                                    ? std::optional{pmr_string{address.street_name1}}
+                                    : std::optional{pmr_string{address.street_name1} + " " + address.street_name2};
+}
+
+// mapping types used by tpcds-dbgen as follows:
+// ds_key_t -> tpcds_key_t
+// int -> int32_t
+// char*, char[], bool, date (ds_key_t as date_id), time (ds_key_t as time_id) -> pmr_string
```

Comment:
Why are we storing date/time as strings, not as ints?

Reply:
Dates are strings in call_center.csv.json etc. - the config files we used already for the file based benchmark.

Reply:
Ok, so it might have been wrong before. My question remains.

Reply:
TPC-DS sometimes uses the type integer and sometimes the type date for dates. Dates are printed like `yyyy-mm-dd` and integers are printed as numbers by dsdgen, so verification gets harder if we use int all the time. We could either use a date type whenever tpc-ds uses date, or modify the verification process to convert `yyyy-mm-dd` to an int before comparing. What do you prefer?

tpcds.sql:
```
create table call_center
(
    cc_call_center_sk         integer               not null,
    cc_call_center_id         char(16)              not null,
    cc_rec_start_date         date                          ,
    cc_rec_end_date           date                          ,
    cc_closed_date_sk         integer                       ,
    cc_open_date_sk           integer                       ,
```

call_center.csv:
```
1|AAAAAAAABAAAAAAA|1998-01-01|||2451086|
```

Reply:
Leaving types for dates as is: date -> string, int -> int.